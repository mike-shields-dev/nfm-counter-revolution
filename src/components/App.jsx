import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import Content from "./Content";
import Disc from "./Disc";

const App = ({ figures }) => {
  const appIdleTimerDurationMillis = 5000;
  const [appIdleTimer, setAppIdleTimer] = useState();
  const baseAnimationDurationMillis = 2400;
  const [index, setIndex] = useState(0);
  const [isAppIdle, setIsAppIdle] = useState(false);

  const incrementIndex = () =>
    setIndex((prev) => (prev < figures.length - 1 ? prev + 1 : 0));

  const startAppIdleTimer = () => {
    setIsAppIdle(false);
    setAppIdleTimer(clearTimeout(appIdleTimer));
  };

  useEffect(() => {
    const doc = document.documentElement;
    doc.style.setProperty(
      "--base-animation-duration",
      `${baseAnimationDurationMillis}ms`
    );
  }, []);

  useEffect(() => {
    if (!(isAppIdle || appIdleTimer)) {
      setAppIdleTimer(
        setTimeout(() => {
          setIsAppIdle(true);
          setAppIdleTimer(clearTimeout(appIdleTimer));
        }, appIdleTimerDurationMillis)
      );
    }
    return () => clearTimeout(appIdleTimer);
  }, [isAppIdle, appIdleTimer]);

  return (
    <div className="App">
      <main>
        <Disc
          {...{
            baseAnimationDurationMillis,
            figures,
            incrementIndex,
            index,
            isAppIdle,
            startAppIdleTimer,
            setIndex,
          }}
        />
        <Content
          {...{
            figures,
            index,
            baseAnimationDurationMillis,
            isAppIdle,
          }}
        />
      </main>
    </div>
  );
};

App.propTypes = {
  figures: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      angle: PropTypes.number,
      text: PropTypes.arrayOf(PropTypes.string),
    })
  ).isRequired,
};

export default App;
