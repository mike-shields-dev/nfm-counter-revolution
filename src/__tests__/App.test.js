import React from "react";
import { render, screen } from "@testing-library/react";
import App from "../components/App";

const validProps = {
  figures: [
    {
      id: 0,
      title: "figure 0",
      angle: 0,
      text: [
        "Figure 0 Paragraph 1",
        "Figure 0 Paragraph 2",
        "Figure 0 Paragraph 3",
      ],
    },
    {
      id: 1,
      title: "figure 1",
      angle: -60,
      text: [
        "Figure 1 Paragraph 1",
        "Figure 1 Paragraph 2",
        "Figure 1 Paragraph 3",
      ],
    },
    {
      id: 2,
      title: "figure 2",
      angle: -120,
      text: [
        "Figure 2 Paragraph 1",
        "Figure 2 Paragraph 2",
        "Figure 2 Paragraph 3",
      ],
    },
  ],
};
const renderApp = () => render(<App {...validProps} />);

describe("App", () => {
  it("matches snapshot", () => {
    const { asFragment } = renderApp();

    expect(asFragment()).toMatchSnapshot();
  });

  beforeEach(() => renderApp());

  it("renders the Content component", () => {
    expect(screen.getByTestId("content")).toBeInTheDocument();
  });

  it("renders the disc component", () => {
    expect(screen.getByTestId("disc")).toBeInTheDocument();
  });
});
