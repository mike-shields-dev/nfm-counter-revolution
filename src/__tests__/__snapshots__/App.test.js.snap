// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <main>
      <svg
        class="disc"
        data-testid="disc"
      >
        <image
          href="/art.png"
        />
        <image
          href="mask0.png"
        />
        <image
          href="mask1.png"
        />
        <image
          href="mask2.png"
        />
        <image
          href="mask3.png"
        />
        <image
          href="mask4.png"
        />
        <image
          href="mask5.png"
        />
      </svg>
      <div
        class="content"
        data-testid="content"
      >
        <div
          class="content__text"
          style="opacity: 0;"
        >
          Call to Action
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
